src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+  }
+
+  /** Gets the sample sheet after adding in any custom sample information tracked by DemuxFastqs. */
+  def sampleSheet(sampleSheet: Path, lane: Option[Int] = None): SampleSheet = {
+    new SampleSheet(
+      SampleSheet(sampleSheet, lane=lane)
+        .map(new CustomSample(_))
+        .toSeq
+    )
+  }
+
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** A custom sample class to support inline sample barcodes for read one and read two. */
+  private[fastq] class CustomSample(sample: Sample)
+    extends Sample(
+      sampleOrdinal      = sample.sampleOrdinal,
+      sampleId           = sample.sampleId,
+      sampleName         = sample.sampleName,
+      libraryId          = sample.libraryId,
+      project            = sample.project,
+      description        = sample.description,
+      lane               = sample.lane,
+      i7IndexBases       = sample.i7IndexBases,
+      i5IndexBases       = sample.i5IndexBases,
+      sampleBarcode      = sample.sampleBarcode,
+      extendedAttributes = sample.extendedAttributes
+    ) {
+    val r1SampleBarcodeBases: Option[String] = extendedAttribute(CustomSampleSheetHeaderNames.R1SampleBarcodeBases)
+    val r2SampleBarcodeBases: Option[String] = extendedAttribute(CustomSampleSheetHeaderNames.R2SampleBarcodeBases)
+
+    /** Returns the sample barcodes in order of sequencing. */
+    override def sampleBarcodeBases: Seq[Option[String]] = {
+      Seq(i7IndexBases, i5IndexBases, r1SampleBarcodeBases, r2SampleBarcodeBases)
+    }
+  }
+
+  /** A custom sample barcode class that stores the reverse complement of the sample barcode. */
+  private[fastq] class CustomSampleBarcode(sampleBarcodes: Seq[String]) extends SampleBarcode(sampleBarcodes) {
+    lazy val reverseComplementedBarcodeBytes: Array[Byte] = {
+      val bytes = barcodeBytes.clone()
+      SequenceUtil.reverseComplement(bytes)
+      bytes
+    }
+  }
+
+
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call. Observed
+    * qualities may not be null. */
+  private[fastq] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
+    var idx = 0
+    var count = 0
+    while (idx < observedBases.length) {
+      val expectedBase = expectedBases(idx)
+      val observedBase = observedBases(idx)
+      if (!SequenceUtil.isNoCall(expectedBase) && !SequenceUtil.basesEqual(observedBase, expectedBase)) {
+        count += 1
+      }
+      idx += 1
+    }
+    count
+  }
+
+  /** Reads multiple FASTQs at the same time, ensuring their read names match along the way.  This is
+    * useful for when we don't have a single interleaved FASTQ. */
+  private[fastq] class ParallelFastqReader(fastq: Seq[Path], fastqs: Seq[Path]*) extends Iterator[Seq[FastqRecord]] {
```

Comment:
Though it's slightly confusing that it's taking many `Seq`s.  Since you can do things like `iterator ++ iterator` in scala, I wonder if this class shouldn't just deal with reading one set of correlated fastqs (i.e. R1, R2, I7, I5) and then outside you could just combine the iterators?